[Research CI/CD] SparseZoo Directory Structure Schemas

dbogunowicz opened the pull request:
- [x] Fix the test in `TestDirectory::test_unzip`

Commit:
initial commit

bfineran commented on src/sparsezoo/objects/model_folder_objects.py:
```diff
@@ -0,0 +1,448 @@
+    def _validate_yaml(self, strict_mode):
+        try:
+            with open(self.path) as file:
+                yaml.load(file, Loader=yaml.FullLoader)
```
this won't work here because recipes with modifiers won't be registered in yaml outside of sparseml

dbogunowicz replied:
Let's discuss it during the sync.

bfineran commented on src/sparsezoo/objects/model_folder_objects.py:
```diff
@@ -0,0 +1,448 @@
+    def __init__(
+        self, name: str, path: Optional[str] = None, url: Optional[str] = None
+    ):
+        """
```
nit: init docstrings go in main class docstring

bfineran commented on src/sparsezoo/objects/model_folder_objects.py:
```diff
@@ -0,0 +1,448 @@
+        :return: File if found, otherwise None
+        """
+        for file in self.files:
+            if isinstance(file, Directory):
```
should also check if directory name matches and return the directory if so

dbogunowicz replied:
Good catch

bfineran reviewed (changes requested):
looks like a good start, discussed let's get these files split out and moved to a directory outside of the soon to be legacy `objects` dir.

we'll also need to build in extra room in these objects to better support file downloads and integration based validation of framework files

Commit:
Finalize TestDirectory::test_unzip test + assert that .get_file() method can also return Directories

dbogunowicz commented on src/sparsezoo/objects/model_folder_objects.py:
```diff
@@ -0,0 +1,452 @@
+         - parent directory of the Directory object (by default)
+         - specified path (if `tar_file_path` provided).
+
+        Note: `path_to_extract` is relative to the working dir!
```
Let's talk about `tar.extractall()`. Either I did not understand properly its purpose or it is a super misleading piece of code.

Commit:
Fix the docstrings (move them from constructor to class definition)

Commit:
Remove 'yaml' support in File class

Commit:
Add protobuf installation to setup.py

bfineran commented on src/sparsezoo/objects/model_folder_objects.py:
```diff
@@ -0,0 +1,442 @@
+        url: Optional[str] = None,
+    ):
+
+        self.files = files
```
can we automatically get the names of the files if given a local directory/tarball?

dbogunowicz replied:
Adding it to the new PR in the other branch.

bfineran commented on src/sparsezoo/objects/model_folder_objects.py:
```diff
@@ -0,0 +1,442 @@
+
+        if path is None:
+            path = self._infer_path_from_files()
+        super().__init__(name=name, path=path, url=url)
```
let's add a property for if the directory is in a tar or untar'd state

dbogunowicz replied:
Adding it to the new PR in the other branch.

bfineran commented on src/sparsezoo/objects/model_folder_objects.py:
```diff
@@ -0,0 +1,442 @@
+
+    # TODO: Add support for model cards (pull integration from
+    #  model.md and specify which files to validate)
+    def validate(self, integration: Optional[str] = None) -> bool:
```
I'm thinking we may want to split out validation for different integrations into subclasses that may be implemented and then can be registered in a dictionary or something similar

dbogunowicz replied:
I propose we add it as a new task for the next sprint. Let's not include this feature into the current review.

bfineran commented on src/sparsezoo/objects/model_folder_objects.py:
```diff
@@ -0,0 +1,442 @@
+
+        super().__init__(files=files, name=name, path=path, url=url)
+
+        self.nested_folders_names = ["checkpoint_(.*)/", "logs/"]
```
nit: let's rename to `valid_nested_folder_patterns`

bfineran commented on src/sparsezoo/objects/model_folder_objects.py:
```diff
@@ -0,0 +1,442 @@
+            (e.g. transformers, YOLOv5 etc.)
+        :return: boolean flag; True if files are valid and no errors arise
+        """
+        # TODO: Assuming for now that all files are loadable
```
should untar first if directory is in tar'd state

bfineran commented on src/sparsezoo/objects/model_folder_objects.py:
```diff
@@ -0,0 +1,442 @@
+                if input_names != list(numpy_dict.keys()):
+                    raise ValueError(
+                        "The keys in npz dictionary do not match "
+                        "the input names of the provided model!"
```
nit: remove `!`, include names of found keys and expected keys with f-string

dbogunowicz replied:
Adding it to the new PR in the other branch.

bfineran commented on src/sparsezoo/objects/model_folder_objects.py:
```diff
@@ -0,0 +1,442 @@
+
+        return True
+
+    def _validate_model(self, model: Optional[onnx.ModelProto]):
```
we need to support both inputs and outputs here since this validates sample inputs and sample outputs (just swapping `model.graph.output` in)

dbogunowicz replied:
Adding it to the new PR in the other branch.

bfineran commented on src/sparsezoo/objects/model_folder_objects.py:
```diff
@@ -0,0 +1,442 @@
+import onnx
+
+from PIL import Image
+from src.sparsezoo.utils.numpy import load_numpy_list
```
```suggestion
from sparsezoo.utils.numpy import load_numpy_list
```

dbogunowicz replied:
Adding it to the new PR in the other branch.

bfineran commented on tests/sparsezoo/objects/test_model_folder_objects.py:
```diff
@@ -0,0 +1,358 @@
+from onnx.helper import make_graph, make_model, make_node, make_tensor_value_info
+
+from PIL import Image
+from src.sparsezoo.objects.model_folder_objects import (
+    Directory,
+    File,
+    FrameworkFiles,
+    NumpyDirectory,
+    SampleOriginals,
+)
+from src.sparsezoo.utils.numpy import load_numpy_list, save_numpy
```
```suggestion
from sparsezoo.objects.model_folder_objects import (
    Directory,
    File,
    FrameworkFiles,
    NumpyDirectory,
    SampleOriginals,
)
from sparsezoo.utils.numpy import load_numpy_list, save_numpy
```

dbogunowicz replied:
Adding it to the new PR in the other branch.

bfineran commented on tests/sparsezoo/objects/test_model_folder_objects.py:
```diff
@@ -0,0 +1,358 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
```
tests look good at a high level - wondering, why create test files on the fly rather than download from the zoo?

bfineran reviewed (changes requested):
great work on these implementations @dbogunowicz, will likely go through further iteration as we build out more of the infra and have a larger variety of sample directories.

To solve the problem of where to put this during development, let's 
1) move all of the "new" code to a temporary `src/sparsezoo/v2/` directory if that works for you
2) let's redirect this PR to `feature/ci-cd-refactor/main` branch where we can merge in all relevant PRs

dbogunowicz commented on setup.py:
```diff
@@ -45,6 +45,8 @@
     "pyyaml>=5.1.0",
     "requests>=2.0.0",
     "tqdm>=4.0.0",
+    "protobuf>=3.19.1",
```
It's a solution for the issue I had in GHA: https://stackoverflow.com/questions/61922334/how-to-solve-attributeerror-module-google-protobuf-descriptor-has-no-attribu

dbogunowicz commented:
@bfineran 

Addressed all the comments above (except for `should untar first if directory is in tar'd state`, this will come in the next iteration) in this commit: https://github.com/neuralmagic/sparsezoo/pull/147/commits/db33e2ab3dd90de5236c87555a0659ec4a80c9ac in #147

bfineran reviewed (approved):
LGTM overall. noting we will need to refactor validation a bit (opened a placeholder ticket in asana) and probably rename some files. Tests look great

dbogunowicz commented:
@bfineran I think it is indeed. Let's close this for now then and continue with #147
